### ISSUE TITLE
DEP-2 — LAMMPS deployment: the MLIAP-first bundle, unified runtime, and trace-based emergency exit

### ISSUE THREAD

**Issue by aacostadiaz** [ACEsuit/mace#1594]
**Depends on:** ARCH-4 (#1563), BKD-2 (#1566), CLI-1 (#1579) · **Blocks:** RET-5 (#1601)

**Prerequisite:** the Phase 0 characterization suite is committed — goldens, fixtures and the single-source tolerance table under `tests/golden/`, characterization tests under `tests/unit/`. References of the form `P0-N` name that work.

**Context:** v1 blocks are born without `@compile_mode`, so neither of today's LAMMPS artifacts can be reproduced: the TorchScript `jit.save` path (`mace/cli/create_lammps_model.py:109-110`, and the `_compiled.model` side artifacts at `mace/cli/run_train.py:1162-1189`) cannot script undecorated blocks, and the ML-IAP pickle path (`create_lammps_model.py:107`) serializes a *live* module that also **converts to cueq by default** (a literal `# Enabling cuequivariance by default. TODO: switch?` at `:90`), baking a vendor backend into a published artifact. Note which of the two is the *default*: `--format` defaults to `libtorch` (`:39-43`), so the shipped default today is the TorchScript one. Until this ticket lands, the energy capability's `--engine=v1` opt-in explicitly excludes LAMMPS export/checkpoint and deployment stays legacy-only (a burn-step in the debt book). This ticket builds the v1 deployment path. The interface to LAMMPS is **ML-IAP** (unconditionally); the v1.0 execution engine behind it is **eager torch + opportunistic per-layer `torch.compile`** (parity-smoked, off-able) — today's mliap flow is already eager, and eager torch runs on device so kokkos/GPU works. `torch.export`/AOTInductor is **not** the v1.0 mechanism: capturing energy+forces means capturing grad-in-forward, which fails on torch 2.11 across all four capture routes and on 2.13 silently returns fake (`FakeTensor`-constant) forces. It is demoted to a **trigger-gated successor**. The deployment artifact is the neutral checkpoint plus a small manifest — no new serialization format, no vendor backend baked in.

**The ghost exchange is a KOKKOS-only capability, and develop already says so out loud.** Every interaction layer past the first needs the node features of the ghost atoms, and the call that supplies them — `forward_exchange` — exists only in LAMMPS's KOKKOS ML-IAP coupling (`src/KOKKOS/mliap_unified_couple_kokkos.pyx`); plain `MLIAPDataPy` has no such method, and conda-forge builds every CPU `lammps` variant with `PKG_KOKKOS=OFF`. Since the fix, `LAMMPS_MLIAP_MACE._check_ghost_exchange_support` (`mace/calculators/lammps_mliap_mace.py:153-175`) **refuses up front** with an actionable `RuntimeError` naming the data class and telling the user to rebuild with `-D PKG_KOKKOS=yes` / `-k on -sf kk` or export a single-layer model — instead of dying on a bare `AttributeError` inside layer two. That refusal is target behaviour, not a defect to remove, and it is why the real-binary test tier exercises a **single-layer** model while the multi-layer path stays in the contract tier.

**Interface & constraints:**

1. **The artifact is a bundle directory, not a new format.** `mace export lammps <checkpoint>` (the `mace_create_lammps_model` successor, `setup.cfg:45`, wired through CLI-1 (#1579)) writes:

```
<name>.mace-deploy/
├── model.safetensors    # neutral weights, byte-identical to the training checkpoint
├── model.json           # ModelConfig + CanonicalSpec sidecar (same as training)
└── deploy.json          # the export manifest — everything deployment-specific
```

2. **The manifest** (`mace_torch.deploy.manifest`); `halo_exchange` is **required, no default** — the presence of the MPI exchange is artifact configuration, never a probe of the compile mode:

```python
@dataclass(frozen=True)
class DeployManifest:
    schema_version: int
    target: Literal["lammps-mliap", "ase", "openmm"]  # extensible
    head: str                     # resolved at export (select_head), never interactive at MD time
    dtype: Precision              # "float64" | "float32" (names, never torch.dtype)
    halo_exchange: bool           # REQUIRED, no default: LAMMPS_MP presence is config, not a compile-mode probe
    backend: str | None = None    # None = resolve at load; a pin is a request, never a bake-in
    compile_regions: bool = True  # opportunistic torch.compile per interaction layer; parity-smoked, off-able
    pad_spec: PadSpec | None = None      # static-shape contract (DATA-3 (#1571)); needed by compile & the adapter
    species: tuple[int, ...] = ()        # atomic numbers, for LAMMPS element mapping
    r_max: float = 0.0
    num_interactions: int = 0            # read by the ghost-exchange precondition check
    receptive_field: float = 0.0         # num_interactions * r_max; consumed when halo_exchange=False
```

   `num_interactions` is in the manifest because the KOKKOS precondition above is a function of it: a bundle with `num_interactions > 1` and `halo_exchange=True` must be refusable **before** the first MD step, from the manifest alone, exactly as legacy now refuses from `self.model.num_interactions`.

3. **The halo-exchange seam (the load-bearing invariant).** Domain decomposition lives **only** on the ML-IAP path: the MPI halo exchange of node features (`LAMMPS_MP`, `mace/tools/utils.py:152`) is applied inside `handle_lammps` between interaction layers (`mace/modules/blocks.py:679-703`; the two `LAMMPS_MP.apply` sites are `:694` and `:703`, with a zero-padded ghost block at `:697-702`) and is today driven by `lammps_class`, set only at `mace/calculators/lammps_mliap_mace.py:221`; the TorchScript artifact never sets it. In v1 there is no TorchScript, so the implicit `torch.jit.is_scripting()` guard (`mace/modules/blocks.py:686`) has **no descendant** — its job is taken over entirely by the `halo_exchange` manifest field. `halo_exchange=True` builds the model with the exchange seam between layers; `halo_exchange=False` requires the consumer to provide a ghost shell of `receptive_field` (the single-rank / native regime). Either way the choice is written down, versioned, and testable. **If presence of the exchange is ever inferred from the compile mode instead of read from the manifest, it silently vanishes from a decomposed run — that is the failure this decision exists to prevent.**

4. **Inference-only, first-order only.** The exchange is a seam between compiled regions (MPI side effects cannot be traced), and its backward is another communication — so there is **no double-backward through it**. LAMMPS is inference-only: energy + per-edge forces are a **first-order** grad (`create_graph=False`), and a backend advertising `supports_double_backward=False` is usable here. **Force training under domain decomposition is not expressible on this path** and is not a goal.

5. **The unified runtime** (`mace_torch.deploy.mliap`) — one class, model built once at setup:

```python
class MACEMLIAPUnified(MLIAPUnified):
    """Successor of LAMMPS_MLIAP_MACE. Loads the bundle, builds the v1 model once
    (backend = manifest.backend or auto; kokkos-aware device selection as today),
    eagerly evaluates energy + per-edge forces (grad wrt LAMMPS-supplied edge vectors,
    first-order, create_graph=False). Refuses at setup, with the legacy message, when
    the bundle needs a ghost exchange the data object cannot provide."""
    def __init__(self, bundle_path: str): ...
    def compute_forces(self, data) -> None: ...   # the MLIAPUnified contract
# LAMMPS input side is unchanged in shape:  pair_style mliap unified <name>.mace-deploy 0
```

The per-edge-force quantity (energy grad w.r.t. edge vectors) is a **first-class observable in `mace_core.observables`** — the same autograd pattern as forces over a different input, with the MLIAP consumer as its one client — **not** a `compute_edge_forces` kwarg threaded through `forward`. A shared loader serves every non-JAX runtime:

```python
# mace_torch.deploy
def load_deployed(path: str, *, device: str = "cpu") -> "DeployedModel":
    """EXTENDS the loader DEP-1 (#1583) already shipped; does not replace it.
    DEP-1 (#1583) handles a plain neutral checkpoint (safetensors + JSON sidecar).
    This ticket teaches the same function to also accept a bundle directory
    and return its manifest alongside the model. One loader for ASE (DEP-1 (#1583)),
    OpenMM (DEP-3 (#1595)) and MLIAP; .compute(graph) -> MACEOutputs."""
```

6. **Env-var runtime config is KEEP, and it is the runtime's, not the artifact's.** `MACELammpsConfig` (`mace/calculators/lammps_mliap_mace.py:21-30`) reads `MACE_TIME`, `MACE_PROFILE`, `MACE_PROFILE_START` (default 5), `MACE_PROFILE_END` (default 10), `MACE_ALLOW_CPU`, `MACE_FORCE_CPU`. These are diagnostics and device-policy overrides for a running MD job — they stay environment variables, they do not migrate into `deploy.json`, and the CPU-tensor guard they gate (`:142-143`, "GPU requested but tensor is on CPU. Set MACE_ALLOW_CPU=true …") keeps its behaviour and its message.

7. **The emergency exit is trace-based, not scriptable.** `torch.jit.script` recurses through the whole module tree, so no wrapper can make undecorated v1 blocks scriptable — that road is a second implementation of every block in the dialect the rewrite exists to remove. The isolable exit is **`torch.jit.trace`**: it records the ops an eager module runs, needs no decorators, and the v1 forward has no data-dependent Python branching on the hot path, so the trace is valid; the padded static shapes are a price deployment already pays.

```python
# mace_torch.deploy.export_adapter  — the ONE whitelisted jit.* site in the v1 tree
def export_torchscript_traced(
    bundle_path: str, pad_spec: "PadSpec", *, check_tolerance: float = 1e-10,
) -> str:
    """EMERGENCY EXIT ONLY (docstring names the trigger that justifies use).
    Builds the model from the bundle (reference backend, eager), traces energy+forces
    at the padded shapes, verifies traced == eager on a golden batch, writes
    <name>-traced.pt for libtorch consumers. Refuses halo_exchange=True (MPI side
    effects cannot be traced)."""
```

It lives in the deploy layer; `nn/`/`models/` cannot import it (import direction is one-way), so it cannot recontaminate the blocks. It carries a debt-book entry from birth: burn-step "delete when trigger T1 promotes torch.export, or when the last libtorch-only consumer is retired."

8. **The five revisit triggers** (filed in the debt book; evaluated at this ticket's kickoff, at Phase 5 exit, and each minor release). The guard is symmetric — T1/T2 stop MACE ignoring torch.export once it works; T3/T4/T5 stop MACE betting on it before it does:

| # | Trigger (testable condition) | Probe | Action when it fires |
|---|---|---|---|
| **T1** promotion | `torch.export.export` captures energy+forces (grad-in-forward, first-order, `create_graph=False`) of the tiny-anchor ScaleShiftMACE on the reference backend, `strict=True`, no graph break, matching eager E/F at fp64 golden tolerance — on **two consecutive** torch minors | `tests/integrations/lammps/test_export_capture.py::test_export_energy_forces` (xfail; its flip *is* the signal). Tries both routes: `autograd.grad`-in-forward and `torch.func.grad` composition — either passing on two consecutive minors flips it | Open the AOTI/ExportedProgram engine ticket: promote torch.export to the shipped MLIAP engine behind the same manifest (`compile_regions` → `exported_program`) |
| **T2** upstream pull | LAMMPS ML-IAP grows an AOTI/ExportedProgram consumption path, **or a CPU `lammps` channel ships `PKG_KOKKOS=ON`** so the real tier can finally run a multi-layer model, or ML-IAP kokkos loses a capability the eager path needs on a supported platform | quarterly upstream review, noted on DEP-2 (#1594)/DEP-3 (#1595); the conda-forge `lammps` build options are part of the review | Re-run the eager-vs-export comparison; promote the multi-layer real-tier test out of the contract tier; T2+T1 make the export engine the default |
| **T3** performance | the v1 eager(+compile) MLIAP path is **>15 % slower** than the legacy TorchScript artifact on **either** committed benchmark case (fp32, NVIDIA fleet): (a) the standard kernel-bound structure, and (b) a **small-subdomain case** (~200–500 atoms, per-rank size of a decomposed run) where host-side overhead is a material fraction of the step — for two consecutive nightlies | the `benchmarks` job in `nightly.yaml`, comparing the two engines while legacy is alive; afterward against the frozen baseline | Spend the `torch.compile` region budget first; if the gap persists, escalate T1 |
| **T4** coverage | a supported platform cannot run Python-embedded LAMMPS (no-Python site policy, or ML-IAP unavailable in a required channel), by an actual user report | issue label `deploy-no-python`, reviewed each release | Ship the §7 trace adapter as a *documented* path (or the exported-program engine if T1 has fired) |
| **T5** adapter health | the trace adapter's parity check fails for a supported model, or torch deprecates `torch.jit.trace` with a removal date | the adapter's own golden test; torch release notes | If T1 fired: delete the adapter. Else: T1's evaluation becomes blocking — deployment must never be left with zero fallback engines |

**Task:**
1. `mace_torch.deploy.manifest`: `DeployManifest` (frozen dataclass, §2) and its schema-versioned round-trip to/from `deploy.json`.
2. Bundle writer behind `mace export lammps` (CLI-1 (#1579) wiring): emit the `<name>.mace-deploy/` directory (weights byte-identical to the training checkpoint), resolving the head at export (successor of `select_head`, `create_lammps_model.py:48-73`, **no** interactive `input()` prompt in the CLI path — legacy prompts at `:63`).
3. `mace_torch.deploy.mliap.MACEMLIAPUnified` and `mace_torch.deploy.load_deployed` (§5); add the per-edge-force observable to `mace_core.observables`; wire the halo exchange as a manifest-configured seam driven by `halo_exchange`, replacing the `is_scripting()` guard; port the ghost-exchange precondition refusal (§Context) so it fires at setup from `num_interactions` + the data object's capabilities, with an equally actionable message.
4. Preserve the six `MACE_*` runtime env vars and the CPU-tensor guard (§6) in the new runtime.
5. Two-tier export contract tests, per `tests/integrations/README.md` and the P0-5 pattern: a **contract tier** without a LAMMPS binary (bundle round-trip, manifest validation, MLIAP-batch numerics vs the ASE calculator on the same structure, the multi-layer refusal — succeeding today's 13 contract tests in `test_export.py`/`test_ghost_parity.py`/`test_mliap_exchange.py` and its hand-written real-vs-ghost harness `_harness.py`), and a **`bin_lammps` nightly tier** running a real single-layer `pair_style mliap unified` against ASE E/F at golden tolerance (succeeding the single test in `test_lammps_real.py`). CI shape is unchanged: a paths-filtered contract job in `ci-integrations.yaml` and a `require-caps: bin_lammps` job in `nightly.yaml`.
6. `export_torchscript_traced` (§7) with its refusals (`halo_exchange=True` rejected), golden self-check, fitness-suite whitelist (the single allowed `jit.*` occurrence), and debt-book entry with burn-step.
7. File the five triggers (§8) in the debt book, and **commit T1's xfail probe** (`test_export_capture.py`) so its flip is the promotion signal.
8. Emit **no** `_compiled.model` side artifact from v1 training (a recorded DROP): deployment artifacts come only from `mace export`; hand the migration story (convert the checkpoint via FM-1 (#1586)/FM-2 (#1587), then `mace export lammps`) to REL-1 (#1603).

**Out of scope:**
- OpenMM, Symmetrix, libmace, and `mace export full-basis-weights` (the reduced→full export for external full-basis reimplementations) — **DEP-3 (#1595)**.
- Deleting the legacy `torch.jit` LAMMPS path (`create_lammps_model.py:109-110`, `run_train.py:1162-1189`) and degrading `tests/integrations/lammps` to golden-export contract tests — **RET-5 (#1601)** (this ticket keeps legacy alive behind the default-legacy launcher).
- `pad_spec` / static-shape semantics — DATA-3 (#1571)'s contract, consumed here.
- JAX deployment — out entirely; mace-jax has no LAMMPS path.
- Force training under domain decomposition — not expressible on this path.

**Acceptance criteria:**
- [ ] `mace export lammps <checkpoint>` writes a `<name>.mace-deploy/` bundle whose `model.safetensors` is byte-identical to the source checkpoint and whose `deploy.json` validates against `DeployManifest` with `halo_exchange` present (no default).
- [ ] Loading a bundle with `backend: None` resolves the backend at load (no vendor pin baked in); a `backend` pin is honoured as a request; there is no cueq-by-default conversion anywhere in the export or load path.
- [ ] The contract-tier test reproduces ASE-calculator energies and forces from `MACEMLIAPUnified` on the same structure at fp64 golden tolerance; the `bin_lammps` nightly tier matches a real single-layer `pair_style mliap unified` run to ASE E/F within golden tolerance.
- [ ] A multi-layer bundle on a data object without `forward_exchange` **refuses at setup** with a message naming the data class and the KOKKOS rebuild — never an `AttributeError` inside layer two.
- [ ] Per-edge forces are computed with `create_graph=False` (first-order); the observable lives in `mace_core.observables`, not as a `forward` kwarg; a backend with `supports_double_backward=False` runs the MLIAP path.
- [ ] With `halo_exchange=True` the model builds the exchange seam between interaction layers; with `False` it consumes `receptive_field`; the choice is read from the manifest and nowhere inferred from the compile mode. No `torch.jit.is_scripting()` survives on the v1 path.
- [ ] `MACE_TIME`, `MACE_PROFILE`, `MACE_PROFILE_START`, `MACE_PROFILE_END`, `MACE_ALLOW_CPU`, `MACE_FORCE_CPU` behave as on develop (defaults 5 / 10 for the profile window), and the CPU-tensor guard still names `MACE_ALLOW_CPU`.
- [ ] `export_torchscript_traced` traces energy+forces at the padded shapes, self-checks traced == eager on a golden batch, refuses `halo_exchange=True`, and is the **only** `jit.*` occurrence the fitness suite allows in the v1 tree.
- [ ] T1's `test_export_capture.py::test_export_energy_forces` is committed and **xfails**, asserting force *values* (realness + golden parity), and tries both the `autograd.grad`-in-forward and `torch.func.grad` routes; the five triggers are recorded in the debt book.
- [ ] v1 training emits no `_compiled.model` artifact.

**Inventory gaps assigned here:**
- [ ] The six MLIAP environment variables — untested on develop; this is where they gain coverage.
- [ ] The multi-layer ghost-exchange refusal — pinned today only by `test_mliap_exchange.py`; it must survive the port as an explicit contract test.

**Verify:**
```bash
python -m pytest tests/integrations/lammps -m "not bin_lammps" -v          # contract tier: bundle round-trip, manifest, MLIAP-vs-ASE numerics, multi-layer refusal
python -m pytest tests/integrations/lammps/test_export_capture.py -v       # T1 probe xfails (its flip is the promotion signal)
python -m pytest packages/mace-torch/tests/deploy/test_export_adapter.py -v # trace self-check + refuses halo_exchange=True
python -m pytest packages/mace-torch/tests/deploy/test_manifest.py -v      # schema-versioned round-trip; halo_exchange required
```
(`bin_lammps` real-tier LAMMPS run is nightly-only, single-layer, `require-caps: bin_lammps,cueq`.)

**Review focus:** that `halo_exchange` is authoritative artifact config and is nowhere re-derived from compile mode (the silent-vanish failure), that the multi-layer refusal fires from the manifest before the first step, that per-edge forces stay first-order (`create_graph=False`) with no double-backward assumed through the exchange, and that the trace adapter is the sole `jit.*` site and cannot be imported by `nn/`/`models/`. Confirm nothing bakes a vendor backend into the published bundle.

---
